components/powersensor/renogy/renogy.go

```diff
@@ -0,0 +1,277 @@
+)
+
+var (
+	globalMu sync.Mutex
+	model    = resource.DefaultModelFamily.WithModel("renogy")
+	readings map[string]interface{}
+)
+
+const (
+	// defaults assume the device is connected via UART serial.
+	pathDefault     = "/dev/serial0"
+	baudDefault     = 9600
+	modbusIDDefault = 1
+
+	solarVoltReg             = 263
+	solarAmpReg              = 264
+	solarWattReg             = 265
+	loadVoltReg              = 260
+	loadAmpReg               = 261
+	loadWattReg              = 262
+	battVoltReg              = 257
+	battChargePctReg         = 256
+	controllerDegCReg        = 259
+	maxSolarTodayWattReg     = 271
+	minSolarTodayWattReg     = 272
+	maxBattTodayVoltReg      = 268
+	minBattTodayVoltReg      = 267
+	maxSolarTodayAmpReg      = 269
+	minSolarTodayAmpReg      = 270
+	chargeTodayWattHrsReg    = 273
+	dischargeTodayWattHrsReg = 274
+	chargeTodayAmpHrsReg     = 275
+	dischargeTodayAmpHrsReg  = 276
+	totalBattOverChargesReg  = 278
+	totalBattFullChargesReg  = 279
+)
+
+// Config is used for converting config attributes.
+type Config struct {
+	resource.TriviallyValidateConfig
+	Path     string `json:"serial_path,omitempty"`
+	Baud     int    `json:"serial_baud_rate,omitempty"`
+	ModbusID byte   `json:"modbus_id,omitempty"`
+}
+
+func init() {
+	resource.RegisterComponent(
+		powersensor.API,
+		model,
+		resource.Registration[powersensor.PowerSensor, *Config]{
+			Constructor: newRenogy,
+		})
+}
+
+func newRenogy(_ context.Context, _ resource.Dependencies, conf resource.Config, logger golog.Logger) (powersensor.PowerSensor, error) {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	if newConf.Path == "" {
+		newConf.Path = pathDefault
+	}
+	if newConf.Baud == 0 {
+		newConf.Baud = baudDefault
+	}
+	if newConf.ModbusID == 0 {
+		newConf.ModbusID = modbusIDDefault
+	}
+
+	return &Renogy{
+		Named:    conf.ResourceName().AsNamed(),
+		path:     newConf.Path,
+		baud:     newConf.Baud,
+		modbusID: newConf.ModbusID,
+	}, nil
+}
+
+// Renogy is a serial charge controller.
+type Renogy struct {
+	resource.Named
+	resource.AlwaysRebuild
+	resource.TriviallyCloseable
+	logger   golog.Logger
+	path     string
+	baud     int
+	modbusID byte
+}
+
+func (r *Renogy) getHandler() *modbus.RTUClientHandler {
+	handler := modbus.NewRTUClientHandler(r.path)
+	handler.BaudRate = r.baud
+	handler.DataBits = 8
+	handler.Parity = "N"
+	handler.StopBits = 1
+	handler.SlaveId = r.modbusID
+	handler.Timeout = 1 * time.Second
+	return handler
+}
+
+// Voltage returns the voltage of the battery and a boolean IsAc.
+func (r *Renogy) Voltage(ctx context.Context, extra map[string]interface{}) (float64, bool, error) {
+	handler := r.getHandler()
+
+	err := handler.Connect()
+	if err != nil {
+		return 0, false, err
+	}
+	client := modbus.NewClient(handler)
+
+	// Read the battery voltage.
+	volts, err := readRegister(client, battVoltReg, 1)
+	if err != nil {
+		return 0, false, err
+	}
+	isAc := false
```

Comment:
prefer to move isAC close to where it is returned, or have it as a global constant in this file